src/mscorlib/src/System/Lazy.cs

```diff
@@ -322,147 +398,116 @@ public T Value
-        private Boxed CreateValue()
+        private sealed class ExecutionAndPublication : LazyInitializer
         {
-            Boxed boxed = null;
-            LazyThreadSafetyMode mode = Mode;
-            if (m_valueFactory != null)
+            internal ExecutionAndPublication(Lazy<T> owner, Func<T> factory) : base(owner, factory) { }
+
+            public override T Value
             {
-                try
+                get
                 {
-                    // check for recursion
-                    if (mode != LazyThreadSafetyMode.PublicationOnly && m_valueFactory == ALREADY_INVOKED_SENTINEL)
-                        throw new InvalidOperationException(Environment.GetResourceString("Lazy_Value_RecursiveCallsToValue"));
-
-                    Func<T> factory = m_valueFactory;
-                    if (mode != LazyThreadSafetyMode.PublicationOnly) // only detect recursion on None and ExecutionAndPublication modes
-                    {
-                        m_valueFactory = ALREADY_INVOKED_SENTINEL;
-                    }
-                    else if (factory == ALREADY_INVOKED_SENTINEL)
+                    lock (this) // we're safe to lock on "this" as object is an private object used by Lazy
                     {
-                        // Another thread raced to successfully invoke the factory.
-                        return null;
+                        return Owner.CreateValueExecutionAndPublication(TakeFactory(), this);
                     }
-                    boxed = new Boxed(factory());
-                }
-                catch (Exception ex)
-                {
-                    if (mode != LazyThreadSafetyMode.PublicationOnly) // don't cache the exception for PublicationOnly mode
-                        m_boxed = new LazyInternalExceptionHolder(ex);
-                    throw;
                 }
             }
-            else
+
+            public override LazyThreadSafetyMode Mode
             {
-                try
-                {
-                    boxed = new Boxed((T)Activator.CreateInstance(typeof(T)));
+                get { return LazyThreadSafetyMode.ExecutionAndPublication; }
+            }
+        }
 
-                }
-                catch (System.MissingMethodException)
-                {
-                    Exception ex = new System.MissingMemberException(Environment.GetResourceString("Lazy_CreateValue_NoParameterlessCtorForT"));
-                    if (mode != LazyThreadSafetyMode.PublicationOnly) // don't cache the exception for PublicationOnly mode
-                        m_boxed = new LazyInternalExceptionHolder(ex);
-                    throw ex;
-                }
+        private sealed class PublicationOnly : LazyInitializer
+        {
+            internal PublicationOnly(Lazy<T> owner, Func<T> factory) : base(owner, factory) { }
+
+            public override T Value
+            {
+                get { return Owner.CreateValuePublicationOnly(Factory, this); }
```

Comment:
No `TakeFactory` here I guess because it's not safe for concurrent calls? How is reentrancy detection handled for `PublicationOnly`